fuel-tx/src/transaction/types/blob.rs

```diff
@@ -0,0 +1,235 @@
+use crate::{
+    transaction::{
+        fee::min_gas,
+        id::PrepareSign,
+        metadata::CommonMetadata,
+        types::chargeable_transaction::{
+            ChargeableMetadata,
+            ChargeableTransaction,
+            UniqueFormatValidityChecks,
+        },
+        Chargeable,
+    },
+    ConsensusParameters,
+    FeeParameters,
+    GasCosts,
+    Input,
+    Output,
+    TransactionRepr,
+    ValidityError,
+};
+use derivative::Derivative;
+use fuel_types::{
+    bytes::WORD_SIZE,
+    canonical::Serialize,
+    BlobId,
+    ChainId,
+    Word,
+};
+
+/// Adds method to `BlobId` to compute the it from blob data.
+pub trait BlobIdExt {
+    /// Computes the `BlobId` from by hashing the given data.
+    fn compute(data: &[u8]) -> Self;
+}
+
+impl BlobIdExt for BlobId {
+    fn compute(data: &[u8]) -> Self {
+        Self::new(*fuel_crypto::Hasher::hash(data))
+    }
+}
+
+pub type Blob = ChargeableTransaction<BlobBody, BlobMetadata>;
+
+#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
+pub struct BlobMetadata;
+
+/// The body of the [`Blob`] transaction.
+#[derive(Clone, Default, Derivative)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(fuel_types::canonical::Deserialize, fuel_types::canonical::Serialize)]
+#[canonical(prefix = TransactionRepr::Blob)]
+#[derivative(Eq, PartialEq, Hash, Debug)]
+pub struct BlobBody {
+    /// Hash of the bytecode. Used both as a unique identifier and to verify the
+    /// bytecode.
+    pub id: BlobId,
+    /// The witness index of the payload.
+    pub witness_index: u16,
+}
+
+impl PrepareSign for BlobBody {
+    fn prepare_sign(&mut self) {}
+}
+
+impl Chargeable for Blob {
+    fn min_gas(&self, gas_costs: &GasCosts, fee: &FeeParameters) -> fuel_asm::Word {
+        min_gas(self, gas_costs, fee)
+    }
+
+    #[inline(always)]
+    fn metered_bytes_size(&self) -> usize {
+        Serialize::size(self)
+    }
+
+    #[inline(always)]
+    fn gas_used_by_metadata(&self, gas_cost: &GasCosts) -> Word {
+        let bytes = Serialize::size(self);
+        // Gas required to calculate the `tx_id`.
+        gas_cost.s256().resolve(bytes as u64)
```

Comment:
And we can also charge for `BlobId` calculation=)

Reply:
ac2776ac